File: loom-workshop-java/src/main/java/nl/theexperts/loomworkshopjava/service/PipelineService.java
package nl.theexperts.loomworkshopjava.service;

import nl.theexperts.loomworkshopjava.repository.DataRecord;
import nl.theexperts.loomworkshopjava.repository.DataRecordRepository;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClientException;
import org.springframework.web.client.RestTemplate;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

@Service
public class PipelineService {

    private final RestTemplate restTemplate;
    private final DataRecordRepository dataRecordRepository;

    public PipelineService(RestTemplate restTemplate, DataRecordRepository dataRecordRepository) {
        this.restTemplate = restTemplate;
        this.dataRecordRepository = dataRecordRepository;
    }

    public List<String> processLargeDataSet(List<String> inputData) {
        List<Future<String>> futures = new ArrayList<>();

        ThreadFactory factory = Thread.ofVirtual().factory();
        try (ExecutorService executor = Executors.newThreadPerTaskExecutor(factory)) {
            for (String data : inputData) {
                Future<String> future = executor.submit(() -> {
                    String externalData = fetchExternalData(data);
                    String dbData = lookupInDatabase(externalData);
                    return intensiveComputationPinned(externalData + " - " + dbData);
                });
                futures.add(future);
            }

            List<String> results = new ArrayList<>();
            for (Future<String> future : futures) {
                results.add(future.get());
            }
            return results;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    private String fetchExternalData(String inputData) {
        String apiUrl = "/" + inputData;
        try {
            restTemplate.getForObject(apiUrl, String.class);
            return inputData;
        } catch (RestClientException e) {
            throw new DataFetchException("Failed to fetch external data.", e);
        }
    }

    private String lookupInDatabase(String data) {
        DataRecord record = dataRecordRepository.findByData(data);
        if (record != null) {
            return record.toString();
        } else {
            throw new DataNotFoundException("Data not found in the database.");
        }
    }

    private String intensiveComputation(String data) {
        // Simulate a heavy computation, e.g., a complex algorithm or just a sleep
        try {
            Thread.sleep(100);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        return new StringBuilder(data).reverse().toString();
    }

    private String intensiveComputationPinned(String data) {
        final Lock lock = new ReentrantLock();
        lock.lock();
        try {
            Thread.sleep(100);
            return new StringBuilder(data).reverse().toString();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } finally {
            lock.unlock();
        }
    }

    static class DataFetchException extends RuntimeException {
        public DataFetchException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    static class DataNotFoundException extends RuntimeException {
        public DataNotFoundException(String message) {
            super(message);
        }
    }
}

File: loom-workshop/src/main/kotlin/nl/theexperts/loomworkshop/service/PipelineService.kt
package nl.theexperts.loomworkshop.service

import nl.theexperts.loomworkshop.repository.DataRecordRepository
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.util.concurrent.Executors
import java.util.concurrent.Future

@Service
class PipelineService(
    private val restTemplate: RestTemplate,
    private val dataRecordRepository: DataRecordRepository
) {

    fun processLargeDataSet(inputData: List<String>): List<String> {
        // Create a list to hold futures of each virtual thread's computation
        val futures = mutableListOf<Future<String>>()

        val factory = Thread.ofVirtual().factory()
        // Create an executor for virtual threads
        val executor = Executors.newThreadPerTaskExecutor(factory)

        // For each data, create a virtual thread and submit the task to the executor
        for (data in inputData) {
            val future = executor.submit<String> {
                val externalData = fetchExternalData(data)
                val dbData = lookupInDatabase(externalData)
                intensiveComputation("$externalData - $dbData")
                // intensiveComputationPinned("$externalData - $dbData")
            }
            futures.add(future)
        }

        // Collect the results from each future (i.e., each virtual thread's computation)
        val results = futures.map { it.get() }

        // Shutdown the executor
        executor.shutdown()

        return results
    }

    private fun fetchExternalData(inputData: String): String {
        val apiUrl = "/$inputData"
        restTemplate.getForObject(apiUrl, String::class.java) ?: throw DataFetchException("Failed to fetch external data.")
        return inputData
    }

    private fun lookupInDatabase(data: String): String {
        val record = dataRecordRepository.findByData(data)
        return record?.data ?: throw DataNotFoundException("Data not found in the database.")
    }

    private fun intensiveComputation(data: String): String {
            Thread.sleep(100)  // Simulate a delay
            return data.reversed()  // Just an example transformation
    }

    private fun intensiveComputationPinned(data: String): String {
        synchronized(this) {
            Thread.sleep(100)  // Simulate a delay
            return data.reversed()  // Just an example transformation
        }
    }
}